### PR TITLE
Refactor DialogService to use model and contributions

### LINKED ISSUE #109056: Cannot use IDialogService in some services due to custom dialog service dependencies
Trying to use `IDialogService` in [ExtensionService](https://github.com/microsoft/vscode/blob/9be5ce5/src/vs/workbench/services/extensions/electron-browser/extensionService.ts#L43) or in its dependencies like [ExtensionManagementService](https://github.com/microsoft/vscode/blob/9be5ce5/src/vs/workbench/services/extensionManagement/common/extensionManagementService.ts#L24) is causing workbench to explode because dialog service is depending on [Layout](https://github.com/microsoft/vscode/blob/9be5ce5/src/vs/workbench/browser/layout.ts#L109) that depends on ExtensionService.

Due to this dependency, DialogService cannot be used in many services. We should allow dialog service to be used by all workbench services and not introduce such a restriction.

### PR DESCRIPTION
**Pull request by sbatten**
This PR fixes #109056

### CODE REVIEW

File: src/vs/platform/dialogs/common/dialogs.ts [hunk 1]
```diff
@@ -166,6 +198,41 @@ export interface IInput {
 	value?: string;
 }
 
+/**
+ * A handler to bring up modal dialogs.
+ *
+ */
```
Comment (bpasero) [inline src/vs/platform/dialogs/common/dialogs.ts:204]:
```suggestion
/**
 * A handler to bring up modal dialogs.
 */
```

File: src/vs/platform/dialogs/common/dialogs.ts [hunk 2]
```diff
@@ -22,6 +22,38 @@ export interface ICheckbox {
...
+
+export interface IInputDialogArgs {
+	severity: Severity;
+	message: string;
+	buttons: string[];
+	inputs: IInput[],
+	options?: IDialogOptions;
+}
+
+export interface IAboutDialogArgs {
```
Comment (bpasero) [inline src/vs/platform/dialogs/common/dialogs.ts:44]: Is this needed when empty?

File: src/vs/workbench/common/dialogs.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	readonly onDidShowDialog: Event<void>;
+
+	readonly dialogs: IDialogViewItem[];
+
+	show(dialog: IDialog): IDialogHandle;
+}
+
+export class DialogsModel extends Disposable implements IDialogsModel {
+	private readonly _dialogs: IDialogViewItem[] = [];
+	get dialogs(): IDialogViewItem[] { return this._dialogs; }
```
Comment (bpasero) [inline src/vs/workbench/common/dialogs.ts:30]:
```suggestion
	readonly dialogs: IDialogViewItem[] = [];
```

File: src/vs/workbench/common/dialogs.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+
```
Comment (bpasero) [inline src/vs/workbench/common/dialogs.ts:32]:
```suggestion

```

File: src/vs/workbench/common/dialogs.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	private readonly _onDidShowDialog = this._register(new Emitter<void>());
+	readonly onDidShowDialog = this._onDidShowDialog.event;
+
+	show(dialog: IDialog): IDialogHandle {
+		let resolver: any;
```
Comment (bpasero) [inline src/vs/workbench/common/dialogs.ts:37]: Sorry, but I do not really tolerate `any` as type.

File: src/vs/workbench/services/dialogs/common/dialogService.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import Severity from 'vs/base/common/severity';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { IConfirmation, IConfirmationResult, IDialogOptions, IDialogService, IInput, IInputResult, IShowResult } from 'vs/platform/dialogs/common/dialogs';
+import { DialogsModel, IDialogsModel } from 'vs/workbench/common/dialogs';
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+
+export class DialogService extends Disposable implements IDialogService {
```
Comment (bpasero) [inline src/vs/workbench/services/dialogs/common/dialogService.ts:12]: I suggest to adopt `async/await` in this file.

File: src/vs/workbench/services/dialogs/common/dialogService.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	_serviceBrand: undefined;
+
+	private _model: IDialogsModel = this._register(new DialogsModel());
+	get model(): IDialogsModel { return this._model; }
```
Comment (bpasero) [inline src/vs/workbench/services/dialogs/common/dialogService.ts:16]:
```suggestion
	readonly model: IDialogsModel = this._register(new DialogsModel());
```

File: src/vs/workbench/services/dialogs/common/dialogService.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	confirm(confirmation: IConfirmation): Promise<IConfirmationResult> {
+		const handle = this.model.show({ confirmArgs: { confirmation } });
+
+		console.log('showing confirm');
```
Comment (bpasero) [inline src/vs/workbench/services/dialogs/common/dialogService.ts:21]: Stray `console.log` in this file

File: src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts [hunk 9]
```diff
@@ -4,7 +4,7 @@
  *--------------------------------------------------------------------------------------------*/
 
 import * as nls from 'vs/nls';
```
Comment (bpasero) [inline src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts:6]: I see a NLS warning here

**Review by bpasero — CHANGES_REQUESTED** [review]
Left some comments inline. Overall I would suggest to make the dialog handlers a core part of the workbench and not add it through contribution mechanism. Similar to notifications, I suggest to add the handlers into the core, e.g.:
* `src/vs/workbench/browser/parts/dialog/...` for the browser handler
* `src/vs/workbench/electron-sandbox/parts/dialog/...` for the native handler

This change would require a way for a dialog handler to register with the workbench I guess. Or maybe we just let them be workbench contributions that listen on the dialog model to participate?

File: src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts (hunk 9 above)
Comment (sbatten) [reply src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts:6]: i don't see this in editor or the watch task

**Comment by sbatten** [comment]
@bpasero I made changes for the inline comments. I'm not sure I understand the benefit/purpose of moving them from workbench contributions which are easily included in the proper `main.ts` to within core, where we then have to instantiate them manually.

File: src/vs/platform/dialogs/common/dialogs.ts [hunk 10]
```diff
@@ -22,6 +22,33 @@ export interface ICheckbox {
 	checked?: boolean;
 }
 
+export interface IConfirmDialogArgs {
+	confirmation: IConfirmation;
+}
+
+export interface IShowDialogArgs {
```
Comment (bpasero) [inline src/vs/platform/dialogs/common/dialogs.ts:29]: I suggest to introduce a common super-type of `IShowDialogArgs` and `IInputDialogArgs` to share properties

File: src/vs/workbench/contrib/dialogs/browser/dialog.contribution.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+		@ILogService logService: ILogService,
+		@ILayoutService layoutService: ILayoutService,
+		@IThemeService themeService: IThemeService,
+		@IKeybindingService keybindingService: IKeybindingService,
+		@IProductService productService: IProductService,
+		@IClipboardService clipboardService: IClipboardService
+	) {
+		this.impl = new HTMLDialogHandler(logService, layoutService, themeService, keybindingService, productService, clipboardService);
+
+		this.model = (this.dialogService as DialogService).model;
```
Comment (bpasero) [inline src/vs/workbench/contrib/dialogs/browser/dialog.contribution.ts:37]: This cast is the consequence of having the handler be in `contrib` and is quite ugly.

File: src/vs/workbench/contrib/dialogs/electron-sandbox/dialog.contribution.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+		@IThemeService themeService: IThemeService,
+		@IKeybindingService keybindingService: IKeybindingService,
+		@IProductService productService: IProductService,
+		@IClipboardService clipboardService: IClipboardService,
+		@INativeHostService nativeHostService: INativeHostService
+	) {
+		this.customImpl = new HTMLDialogHandler(logService, layoutService, themeService, keybindingService, productService, clipboardService);
+		this.nativeImpl = new NativeDialogHandler(logService, nativeHostService, productService, clipboardService);
+
+		this.model = (this.dialogService as DialogService).model;
```
Comment (bpasero) [inline src/vs/workbench/contrib/dialogs/electron-sandbox/dialog.contribution.ts:44]: Same

File: src/vs/workbench/services/dialogs/common/dialogService.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+
+export class DialogService extends Disposable implements IDialogService {
+	_serviceBrand: undefined;
+
+	readonly model: IDialogsModel = this._register(new DialogsModel());
+
+	async confirm(confirmation: IConfirmation): Promise<IConfirmationResult> {
+		const handle = this.model.show({ confirmArgs: { confirmation } });
+		return await handle.result as IConfirmationResult;
```
Comment (bpasero) [inline src/vs/workbench/services/dialogs/common/dialogService.ts:19]: Why do we need to `as` cast here and below?

**Review by bpasero — COMMENTED** [review]
@sbatten I think the workbench should provide dialogs out of core similar to notifications. If you ship the core workbench without contrib, dialogs will not work otherwise. I can do this change once you merged to `master` too, as you like.

Left some more comments.

File: src/vs/workbench/services/dialogs/common/dialogService.ts (hunk 13 above)
Comment (sbatten) [reply src/vs/workbench/services/dialogs/common/dialogService.ts:19]: the result is a union type but the return type can only be one

File: src/vs/platform/dialogs/common/dialogs.ts (hunk 10 above)
Comment (sbatten) [reply src/vs/platform/dialogs/common/dialogs.ts:29]: made inputargs extend showargs following the pattern of result

**Comment by sbatten** [comment]
@bpasero if you already have an idea where you want to move the files to be part of the core workbench, feel free to do so. please cc me on the commit so I can understand further the changes.

File: src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts (hunk 9 above)
Comment (bpasero) [reply src/vs/workbench/contrib/dialogs/browser/dialogHandler.ts:6]: @sbatten please make sure you have setup vscode properly to see these warnings, todays build is broken: https://dev.azure.com/vscode/VSCode/_build/results?buildId=57514&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=3367cfe7-6aa5-5a88-c888-e6814d74f6eb&l=25

**Pull request merged**